arangod/Pregel/Algos/AIR/Greenspun/Primitives.cpp

```diff
@@ -0,0 +1,1264 @@
+  } else {
+    return EvalError("key is neither array nor string");
+  }
+
+  return {};
+}
+
+EvalResult Prim_Lambda(Machine& ctx, VPackSlice const paramsList, VPackBuilder& result) {
+  VPackArrayIterator paramIterator(paramsList);
+  if (!paramIterator.valid()) {
+    return EvalError(
+        "lambda requires two arguments: a list of argument names and a body");
+  }
+
+  auto captures = *paramIterator++;
+  if (captures.isArray()) {
+    for (auto&& name : VPackArrayIterator(captures)) {
+      if (!name.isString()) {
+        return EvalError("in capture list: expected name, found: " + name.toJson());
+      }
+    }
+  }
+
+  auto params = *paramIterator++;
+  if (params.isArray()) {
+    for (auto&& name : VPackArrayIterator(params)) {
+      if (!name.isString()) {
+        return EvalError("in parameter list: expected name, found: " + name.toJson());
+      }
+    }
+  }
+
+  if (!paramIterator.valid()) {
+    return EvalError("missing body");
+  }
+
+  auto body = *paramIterator++;
+  if (paramIterator.valid()) {
+    return EvalError("to many arguments to lambda constructor");
+  }
+
+  {
+    VPackObjectBuilder ob(&result);
+    result.add("_params", params);
+    result.add("_call", body);
+    {
+      VPackObjectBuilder cob(&result, "_captures");
+      for (auto&& name : VPackArrayIterator(captures)) {
+        result.add(name);
+        if (auto res = ctx.getVariable(name.copyString(), result); res.fail()) {
+          return res;
+        }
+      }
+    }
+  }
+  return {};
+}
+
+EvalResult EvaluateApply(Machine& ctx, VPackSlice const functionSlice,
+                         VPackArrayIterator paramIterator, VPackBuilder& result,
+                         bool isEvaluateParameter);
+
+EvalResult Prim_Apply(Machine& ctx, VPackSlice const paramsList, VPackBuilder& result) {
+  if (!paramsList.isArray() || paramsList.length() != 2) {
+    return EvalError(
+        "expected one function argument on one list of parameters");
+  }
+
+  auto functionSlice = paramsList.at(0);
+  auto parameters = paramsList.at(1);
+  if (!parameters.isArray()) {
+    return EvalError("expected list of parameters, found: " + parameters.toJson());
+  }
+
+  return EvaluateApply(ctx, functionSlice, VPackArrayIterator(parameters), result, false);
+}
+
+EvalResult Prim_Identity(Machine& ctx, VPackSlice const paramsList, VPackBuilder& result) {
+  if (!paramsList.isArray() || paramsList.length() != 1) {
+    return EvalError("expecting a single argument");
+  }
+
+  result.add(paramsList.at(0));
+  return {};
+}
+
+EvalResult Prim_Map(Machine& ctx, VPackSlice const paramsList, VPackBuilder& result) {
+  if (!paramsList.isArray() || paramsList.length() != 2) {
+    return EvalError("expecting to arguments, a function and a list");
+  }
+
+  auto functionSlice = paramsList.at(0);
+  auto list = paramsList.at(1);
+
+  if (list.isArray()) {
+    VPackArrayBuilder ab(&result);
+    for (VPackArrayIterator iter(list); iter.valid(); iter++) {
```

Comment:
VPackArrayIterator is not completely trivial. We should use preincrement on it.
```suggestion
    for (VPackArrayIterator iter(list); iter.valid(); ++iter) {
```